### PR TITLE
Update mlflow.spark.log_model() to accept descendants of pyspark.Model

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?
 
The mlflow.spark.log_model() method defines a spark_model parameter. This parameter corresponds to the PySpark Pipeline that the caller is persisting in MLflow format. Currently, the spark_model parameter must be of type pyspark.ml.PipelineModel; the mlflow.spark._validate_model() method throws an exception if spark_model is not of type PipelineModel.

However, a PipelineModel is just a collection of PySpark stages, and each stage is a descendant of pyspark.ml.Transformer. We should be able to handle any descendant of pyspark.ml.Model (a descendant of Transformer) because those generate predictions, and we can wrap these under the hood in a PipelineModel that we then save. We cannot support arbitrary Transformers because many of them generate "features" and not predictions which breaks some other assumptions in mlflow.spark

This commit extends the set of supported types for the spark_model parameter of mlflow.spark.log_model() to include any descendant of pyspark.ml.Model. We convert it to a PipelineModel to use a consistent save/load implementation before serializing
 
## How is this patch tested?
 
Added new tests for different types of pyspark models for `log_model` and `save_model`
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
mlflow.spark.log_model and mlflow.spark.save_model can now serialize any descendant of pyspark.ml.Model which implements MLReadable and MLWritable.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [x] Docs
- [x] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/spark.py [hunk 1]
```diff
@@ -250,11 +253,14 @@ def _save_model_metadata(dst_dir, spark_model, mlflow_model, sample_input, conda
...
-    if not isinstance(spark_model, PipelineModel):
-        raise MlflowException("Not a PipelineModel. SparkML can only save PipelineModels.",
-                              INVALID_PARAMETER_VALUE)
+    from pyspark.ml import Model
+    from pyspark.ml.util import MLReadable, MLWritable
+    if not isinstance(spark_model, Model) \
+            or not isinstance(spark_model, MLReadable) \
+            or not isinstance(spark_model, MLWritable):
+        raise MlflowException(
+                "Not a descendant of pyspark.Model. SparkML can only save descendants of pyspark.Model that implement MLWritable and MLReadable.",
```
Comment (dbczumar) [inline mlflow/spark.py]: Nit: We're always including `Not a descendant of pyspark.Model` in the error text, even if the  `isinstance(spark_model, Model)` check passes. Maybe we should say something like `Cannot serialize model input. SparkML can only save...`.

File: tests/spark/test_spark_model_export.py [hunk 2]
```diff
@@ -155,6 +198,21 @@ def test_model_export(spark_model_iris, model_path, spark_custom_env):
...
+    preds_df = reloaded_model.transform(spark_model_estimator.spark_df)
+    preds = [x.prediction for x in preds_df.select("prediction").collect()]
+    assert spark_model_estimator.predictions == preds
+
+
+@pytest.mark.large
+def test_transformer_model_export(spark_model_transformer, model_path, spark_custom_env):
+    with pytest.raises(MlflowException):
+        sparkm.save_model(spark_model_transformer.model, path=model_path, conda_env=spark_custom_env)
+
```
Comment (dbczumar) [inline tests/spark/test_spark_model_export.py]: Nit: Can we check that the error message contains some of the expected text, as we do in `test_sparkml_model_log_invalid_args`?

File: tests/spark/test_spark_model_export.py [hunk 3]
```diff
@@ -92,6 +92,49 @@ def spark_model_iris(spark_context):
                               predictions=preds)
 
 
+@pytest.fixture(scope="session")
+def spark_model_transformer(spark_context):
+    iris = datasets.load_iris()
```
Comment (dbczumar) [inline tests/spark/test_spark_model_export.py]: Nit: We can probably DRY this out a bit. Consider defining a fixture that creates pandas and spark representations of the iris data (the fixture can return a tuple).

**Review by dbczumar — CHANGES_REQUESTED** [review]
Looks great in general! Left a few nits. Additionally, we should update the parameter documentation for `spark_model` to reflect the new requirements (`spark_model` can now be any `pyspark.ml.Model` instance implementing the `MLWritable` and `MLReadable` interfaces).

File: mlflow/spark.py (hunk 1 above)
Comment (ankitmathur-db) [reply mlflow/spark.py]: Yeah, good point. Adjusting this and the test

**Review by dbczumar — APPROVED** [review]
LGTM! Nice work @ankitmathur-db!

**Pull request merged**